src/pytorch_lightning/strategies/ddp.py

```diff
@@ -296,39 +304,44 @@ def optimizer_step(
     def configure_ddp(self) -> None:
         log.detail(f"{self.__class__.__name__}: configuring DistributedDataParallel")
         self.pre_configure_ddp()
+        assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
         self.model = self._setup_model(LightningDistributedModule(self.model))
         self._register_ddp_hooks()
 
-    def determine_ddp_device_ids(self):
+    def determine_ddp_device_ids(self) -> Optional[List[int]]:
         if self.root_device.type == "cpu":
             return None
         return [self.root_device.index]
 
-    def barrier(self, *args, **kwargs) -> None:
+    def barrier(self, *args: Any, **kwargs: Any) -> None:
         if not distributed_available():
             return
         if torch.distributed.get_backend() == "nccl":
             torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
         else:
             torch.distributed.barrier()
 
-    def broadcast(self, obj: object, src: int = 0) -> object:
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
         obj = [obj]
         if self.global_rank != src:
-            obj = [None]
+            obj = [None]  # type: ignore[list-item]
         torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
         return obj[0]
 
     def pre_backward(self, closure_loss: Tensor) -> None:
         """Run before precision plugin executes backward."""
-        if not self.lightning_module.automatic_optimization:
+        if isinstance(self.lightning_module, pl.LightningModule) and not self.lightning_module.automatic_optimization:
+            assert isinstance(self.model, DistributedDataParallel)
             prepare_for_backward(self.model, closure_loss)
 
-    def model_to_device(self):
+    def model_to_device(self) -> None:
         log.detail(f"{self.__class__.__name__}: moving model to device [{self.root_device}]...")
+        assert self.model is not None
         self.model.to(self.root_device)
 
-    def reduce(self, tensor, group: Optional[Any] = None, reduce_op: Union[ReduceOp, str] = "mean") -> Tensor:
+    def reduce(
+        self, tensor: Tensor, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = "mean"
```

Comment:
```suggestion
        self, tensor: Tensor, group: Optional[Any] = None, reduce_op: Union[ReduceOp, str] = "mean"
```
since the default is not `None`? just guessing

Reply:
In `reduce()`, parameter reduce_op in `sync_ddp_if_available` is hinted `Optional[Union[ReduceOp, str]]` and `reduce()` in class `Strategy` too, so i thought it's Optional.
```Python
def sync_ddp_if_available(
    result: Tensor, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = None
) -> Tensor:
```
```Python
@abstractmethod
def reduce(
    self,
    tensor: Union[Tensor, Any],
    group: Optional[Any] = None,
    reduce_op: Optional[Union[ReduceOp, str]] = "mean",
) -> Union[Tensor, Any]:
```

Reply:
Yes it is optional for backward compat when we at some point change to the "mean" default value. There are some inconsistencies here, for example in the tpu strategy, which we can standardize in a separate PR.

Reply:
@rohitgr7 Good with you if we break support to set it None in a different PR?

Reply:
yeah. sure!